lib/inventory/controller.go

```diff
@@ -1261,59 +1274,53 @@ func (c *Controller) handleAgentMetadata(handle *upstreamHandle, m *proto.Upstre
 }
 
 func (c *Controller) keepAliveAppServer(handle *upstreamHandle, now time.Time, name resourceKey) error {
-	srv, ok := handle.appServers[name]
-	if !ok {
+	srv := handle.appServers[name]
+	if srv == nil {
 		handle.appKeepAliveDelay.Remove(name)
 		return trace.Errorf("desync between app server hb registry and keepalive delay (this is a bug)")
 	}
 
-	if srv.lease != nil {
-		lease := *srv.lease
-		lease.Expires = now.Add(c.serverTTL).UTC()
-		if err := c.auth.KeepAliveServer(c.closeContext, lease); err != nil {
-			c.testEvent(appKeepAliveErr)
-
-			srv.keepAliveErrs++
-			handle.appServers[name] = srv
-			shouldRemove := srv.keepAliveErrs > c.maxKeepAliveErrs
-			slog.WarnContext(c.closeContext, "Failed to keep alive app server",
-				"server_id", handle.Hello().ServerID,
-				"error", err,
-				"error_count", srv.keepAliveErrs,
-				"should_remove", shouldRemove,
-			)
-
-			if shouldRemove {
-				c.testEvent(appKeepAliveDel)
-				c.onDisconnectFunc(constants.KeepAliveApp, 1)
-				if c.appHBVariableDuration != nil {
-					c.appHBVariableDuration.Dec()
-				}
-				delete(handle.appServers, name)
-				handle.appKeepAliveDelay.Remove(name)
-			}
+	srv.resource.SetExpiry(now.Add(c.serverTTL).UTC())
+	if _, err := c.auth.UnconditionalUpdateApplicationServer(c.closeContext, srv.resource); err == nil {
```

Comment:
Why does the internal keepAliveAppServer controller flow need to send the full srv.resource object over the wire and perform a full object write?

Previously, the KeepAliveServer approach was used, which is a much lighter write call – it doesn’t do any resource JSON marshaling or a full object write, but simply updates the record expiration time.


So comapring old flow with very have `KeepAliveServer` vs  UnconditionalUpdateApplicationServer that tries to update expiration time via full object write at first glance don't make seanse. 


Even with reducing the write calls with resoucse  been change  from handleAppServerHB triggered via agent heartbeat we have switch from very light  ` c.auth.KeepAliveServer` to the very heavy `.auth.UnconditionalUpdateApplicationServer` write call.

Reply:
> the full upsert ensured that the value in storage was eventually going to be set to the value that the agent intended


Does this is reallya  issue ?  how in the current flow  can change witghout recing the app_sever HB ? 

Where the handleAppServerHB handler upset the resourse (after your change upsert on the equa.Difference) 
```
= c.auth.UpsertApplicationServer(c.closeContext, srv.resource); err == nil {
```

> but if we skip the full upsert we're not going to notice if something else writes on the app server.

How the app_sever can change and this change not beeing propagated in HB handleAppServerHB agent `app_sever` heartbeat mechanism ?